core/client/hammer/hammer.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package hammer sends multiple requests to Key Transparency at the same time.
+package hammer
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/google/tink/go/tink"
+	"google.golang.org/grpc"
+
+	"github.com/google/keytransparency/core/authentication"
+	"github.com/google/keytransparency/core/client"
+
+	tpb "github.com/google/keytransparency/core/api/type/type_go_proto"
+	pb "github.com/google/keytransparency/core/api/v1/keytransparency_go_proto"
+	tinkpb "github.com/google/tink/proto/tink_go_proto"
+)
+
+// CustomDial returns a grpc connection.
+var CustomDial func(ctx context.Context, addr string, opts ...grpc.DialOption) (pb.KeyTransparencyClient, error)
+
+// Hammer represents a single run of the hammer.
+type Hammer struct {
+	workers int
+
+	timeout time.Duration
+	ktAddr  string
+	appID   string
+	config  *pb.Domain
+
+	signers        []*tink.KeysetHandle
+	authorizedKeys *tinkpb.Keyset
+}
+
+// New returns a new hammer job
+func New(ctx context.Context, ktAddr, domainID string, timeout time.Duration, keyset *tink.KeysetHandle) (*Hammer, error) {
+	client, err := CustomDial(ctx, ktAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	config, err := client.GetDomain(ctx, &pb.GetDomainRequest{DomainId: domainID})
+	if err != nil {
+		return nil, err
+	}
+
+	authorizedKeys, err := keyset.GetPublicKeysetHandle()
+	if err != nil {
+		return nil, fmt.Errorf("keyset.GetPublicKeysetHandle() failed: %v", err)
+	}
+
+	return &Hammer{
+		timeout: timeout,
+		ktAddr:  ktAddr,
+		appID:   fmt.Sprintf("hammer_%v", time.Now().Format("2006-01-02/15:04:05")),
+		config:  config,
+
+		signers:        []*tink.KeysetHandle{keyset},
+		authorizedKeys: authorizedKeys.Keyset(),
+	}, nil
+}
+
+// Run runs operationCount with up to maxWorkers over ramp duration.
+func (h *Hammer) Run(ctx context.Context, operationCount, maxWorkers int, ramp time.Duration) {
```

Comment:
Maybe make this whole thing (down to `worker`) `ctx`-cancellable?

Reply:
You use `case <-ctx.Done():` clause below in `gen` function to make it early cancellable, right? I was suggesting to maybe do the same in `launchWorkers` and `printStats`. But after second thought I see that those will actually terminate after `gen` gets cancelled, as they both depend on it through channels.
So you can ignore this comment.